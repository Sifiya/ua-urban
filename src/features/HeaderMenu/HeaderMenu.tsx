'use client';
import React, { useMemo } from 'react';
import { usePathname } from 'next/navigation';
import { useProfile } from '@/hooks/useProfile';
import { useQueryClient } from '@tanstack/react-query';
import { signOut } from '@/app/api/auth.api';

import { 
  NavigationMenu,
  NavigationMenuItem,
  NavigationMenuLink,
  NavigationMenuList,
  navigationMenuTriggerStyle,
} from '@/components/ui/navigation-menu';
import Link from 'next/link';
import { SignInForm, SignUpForm } from '../Auth';
import { Button } from '@/components/ui/button';


const MENU_LINKS = [
  { href: '/', label: 'На головну' },
  { href: '/add', label: 'Додати слово' },
];

export const HeaderMenu = () => {
  const { isAuthenticated, email } = useProfile();

  const filteredLinks = useMemo(() => MENU_LINKS.filter((link) => {
    return link.href !== '/add' || isAuthenticated;
  }), [isAuthenticated]);

  const queryClient = useQueryClient();
  const pathname = usePathname();
  const onSignOut = async () => {
    const { error } = await signOut();
    if (error) {
      console.log(error);
    } else {
      queryClient.invalidateQueries({
        queryKey: ['profile']
      });
    }
  }

  return (
    <div className="w-full flex flex-row justify-between px-5">
      <NavigationMenu>
        <NavigationMenuList>
          {filteredLinks.map(({ href, label }) => (
            <NavigationMenuItem key={`${href.replace('/','')}`}>
              <Link 
                href={href} 
                passHref
                legacyBehavior>
                <NavigationMenuLink active={pathname === href} className={navigationMenuTriggerStyle()} >
                  {label}
                </NavigationMenuLink>
              </Link>
            </NavigationMenuItem>
          ))}
        </NavigationMenuList>
      </NavigationMenu>

      {!isAuthenticated && (
        <div className="flex items-center gap-3">
          <SignInForm />
          <SignUpForm />
        </div>
      )}
      {isAuthenticated && (
        <div className="flex items-center gap-3">
          <span>{email}</span>
          <Button variant="outline" onClick={onSignOut}>Вийти</Button>
        </div>
      )}
    </div>
  );
};
