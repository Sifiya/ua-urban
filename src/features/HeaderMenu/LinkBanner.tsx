'use server';
import React from 'react';
import { headers } from 'next/headers';
import Link from 'next/link';
import { FaLink } from 'react-icons/fa';

interface LinkBannerProps {}

const OFFICIAL_HOST = 'www.urban.in.ua';

export const LinkBanner = ({}: LinkBannerProps) => {
  const headersList = headers();
  const host = headersList.get('host');

  if (!host || OFFICIAL_HOST.includes(host)) {
    return null;
  }

  return (
   
    <Link 
      href={`https://${OFFICIAL_HOST}`}
      className="bg-primary text-primary-foreground text-sm w-full p-3 flex items-center justify-center hover:opacity-90 transition-opacity"
    >
        <FaLink className="mr-2" />
        <span>Сайт переїхав на</span>
        <span className="underline ml-1 text-secondary">{OFFICIAL_HOST}</span>
        <FaLink className="ml-2" />
    </Link>
  );
};
