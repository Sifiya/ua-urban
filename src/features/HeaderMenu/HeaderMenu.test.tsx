import React from 'react';
import { render, screen } from '@testing-library/react';
import { userEvent, UserEvent } from '@testing-library/user-event';
import { HeaderMenu } from './HeaderMenu';

let user: UserEvent;

const useMediaQueryMock = jest.fn();
jest.mock('@/hooks/useMediaQuery', () => ({ useMediaQuery: jest.fn().mockImplementation(() => useMediaQueryMock()) }));
const useProfileMock = jest.fn();
jest.mock('@/hooks/useProfile', () => ({ useProfile: jest.fn().mockImplementation(() => useProfileMock() ) }));
const useQueryClientMock = jest.fn();
jest.mock('@tanstack/react-query', () => ({
  useQueryClient: jest.fn().mockImplementation(() => useQueryClientMock())
}));
const signOutMock = jest.fn();
jest.mock('@/app/api/auth.api', () => ({
  signOut: jest.fn().mockImplementation(() => signOutMock())
}));

describe('HeaderMenu', () => {
  beforeEach(() => {
    user = userEvent.setup({ delay: null });
    useProfileMock.mockReturnValue({ isAuthenticated: false });
  });

  afterEach(() => {
    useProfileMock.mockReset();
    signOutMock.mockClear();
  });

  describe('on mobile', () => {
    const getHambugerButton = () => screen.getByRole('button', { name: 'Відкрити меню' });
    const clickHambugerButton = async () => user.click(getHambugerButton());

    beforeEach(() => {
      useMediaQueryMock.mockReturnValue(false);
    });

    test('renders dropdown menu', () => {
      const wrapper = render(<HeaderMenu />);

      // Assert
      expect(getHambugerButton()).toBeInTheDocument();
      expect(screen.queryByRole('navigation')).not.toBeInTheDocument();
      expect(wrapper).toMatchSnapshot();
    });

    test('shows full menu after click on burger icon', async () => {
      render(<HeaderMenu />);

      // Act
      await clickHambugerButton();

      // Assert
      expect(await screen.findByRole('group')).toBeInTheDocument();
      expect(screen.getByRole('link', { name: 'На головну' })).toBeInTheDocument();
    });

    test('if user is not authenticated, should show according buttons', async () => {
      useProfileMock.mockReturnValue({ isAuthenticated: false });
      render(<HeaderMenu />);

      // Act
      await clickHambugerButton();

      // Assert
      await screen.findByRole('group');
      expect(screen.queryByRole('link', { name: 'Додати слово' })).not.toBeInTheDocument();
      expect(screen.getByText('Вхід')).toBeInTheDocument();
      expect(screen.getByText('Реєстрація')).toBeInTheDocument();
    });

    test('if user is authenticated, should show according buttons', async () => {
      useProfileMock.mockReset();
      const email = 'fakemail@mail.com';
      useProfileMock.mockReturnValue({ isAuthenticated: true, email });
      render(<HeaderMenu />);

      // Act
      await clickHambugerButton();

      // Assert
      await screen.findByRole('group');
      expect(screen.getByRole('link', { name: 'Додати слово' })).toBeInTheDocument();
      expect(screen.getByText('Вийти')).toBeInTheDocument();
      expect(screen.getByText(email)).toBeInTheDocument();
      expect(screen.queryByText('Вхід')).not.toBeInTheDocument();
      expect(screen.queryByText('Реєстрація')).not.toBeInTheDocument();
    });
  });

  describe('on desktop', () => {
    beforeEach(() => {
      useMediaQueryMock.mockReturnValue(true);
    });

    test('renders navigation menu', () => {
      render(<HeaderMenu />);

      // Assert
      expect(screen.queryByRole('button', { name: 'Відкрити меню' })).not.toBeInTheDocument();
      expect(screen.getByRole('navigation')).toBeInTheDocument();
    });

    test('if user is not authenticated, should show according buttons', async () => {
      useProfileMock.mockReturnValue({ isAuthenticated: false });
      render(<HeaderMenu />);

      // Assert
      expect(screen.queryByRole('link', { name: 'Додати слово' })).not.toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Вхід' })).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Реєстрація' })).toBeInTheDocument();
    });

    test('if user is authenticated, should show according buttons', async () => {
      useProfileMock.mockReset();
      const email = 'fakemail@mail.com';
      useProfileMock.mockReturnValue({ isAuthenticated: true, email });
      render(<HeaderMenu />);

      // Assert
      expect(screen.getByRole('link', { name: 'Додати слово' })).toBeInTheDocument();
      expect(screen.getByText('Вийти')).toBeInTheDocument();
      expect(screen.getByText(email)).toBeInTheDocument();
      expect(screen.queryByText('Вхід')).not.toBeInTheDocument();
      expect(screen.queryByText('Реєстрація')).not.toBeInTheDocument();
    });
  });

  describe('sign out', () => {
    test('should sign out user', async () => {
      useMediaQueryMock.mockReturnValue(true);
      let invalidateQueriesMock = jest.fn();
      useQueryClientMock.mockReturnValue({ invalidateQueries: invalidateQueriesMock });
      useProfileMock.mockReturnValue({ isAuthenticated: true, email: 'fakemail@mail.com' });
      signOutMock.mockResolvedValue({});
      render(<HeaderMenu />);

      // Act
      await user.click(screen.getByRole('button', { name: 'Вийти' }));
      expect(signOutMock).toHaveBeenCalledTimes(1);
      expect(invalidateQueriesMock).toHaveBeenCalledWith({ queryKey: ['profile'] });
    });

    test('should log error if sign out failed', async () => {
      useMediaQueryMock.mockReturnValue(true);
      let invalidateQueriesMock = jest.fn();
      useQueryClientMock.mockReturnValue({ invalidateQueries: invalidateQueriesMock });
      useProfileMock.mockReturnValue({ isAuthenticated: true, email: 'fakemail@mail.com' });

      const error = new Error('Some error');
      signOutMock.mockResolvedValue({ error });

      render(<HeaderMenu />);
      await user.click(screen.getByRole('button', { name: 'Вийти' }));
      expect(signOutMock).toHaveBeenCalledTimes(1);
      expect(invalidateQueriesMock).not.toHaveBeenCalled();
    });
  });
});

