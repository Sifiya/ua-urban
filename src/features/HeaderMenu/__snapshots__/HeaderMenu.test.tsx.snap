// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderMenu on mobile renders dropdown menu 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        aria-expanded="false"
        aria-haspopup="menu"
        data-state="closed"
        id="radix-:r0:"
        type="button"
      >
        <svg
          aria-label="Відкрити меню"
          fill="currentColor"
          height="24"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
          />
        </svg>
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      aria-expanded="false"
      aria-haspopup="menu"
      data-state="closed"
      id="radix-:r0:"
      type="button"
    >
      <svg
        aria-label="Відкрити меню"
        fill="currentColor"
        height="24"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 448 512"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
        />
      </svg>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
